cw-orch-daemon/src/tx_broadcaster.rs

```diff
@@ -0,0 +1,226 @@
+use std::time::Duration;
+
+use cosmrs::proto::cosmos::base::abci::v1beta1::TxResponse;
+use secp256k1::All;
+
+use crate::{
+    queriers::{DaemonQuerier, Node},
+    sender::Sender,
+    DaemonError, TxBuilder,
+};
+
+pub type StrategyAction =
+    fn(&mut TxBuilder, &Result<TxResponse, DaemonError>) -> Result<(), DaemonError>;
+
+pub struct RetryStrategy {
+    /// This correspond to errors caught when submitting transactions
+    pub broadcast_condition: fn(&TxResponse) -> bool,
+    /// This correspond to errors caught when simulating the transaction
+    pub simulation_condition: fn(&DaemonError) -> bool,
+    pub action: Option<StrategyAction>,
+    pub max_retries: BroadcastRetry,
+    pub current_retries: u64,
+    pub reason: String,
+}
+
+impl RetryStrategy {
+    pub fn new(
+        broadcast_condition: fn(&TxResponse) -> bool,
+        simulation_condition: fn(&DaemonError) -> bool,
+        action: Option<StrategyAction>,
+        max_retries: BroadcastRetry,
+        reason: String,
+    ) -> Self {
+        RetryStrategy {
+            broadcast_condition,
+            simulation_condition,
+            action,
+            max_retries,
+            current_retries: 0,
+            reason,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct TxBroadcaster {
+    strategies: Vec<RetryStrategy>,
+}
+
+pub enum BroadcastRetry {
+    Infinite,
+    Finite(u64),
+}
+
+impl TxBroadcaster {
+    pub fn add_strategy(mut self, s: RetryStrategy) -> Self {
+        self.strategies.push(s);
+        self
+    }
+
+    // We can't make async recursions easily because wallet is not `Sync`, so we have this shit function structure
+    pub async fn broadcast(
+        mut self,
+        mut tx_builder: TxBuilder,
+        wallet: &Sender<All>,
+    ) -> Result<TxResponse, DaemonError> {
+        let mut tx_retry = true;
+
+        // We try and broadcast once
+        let mut tx_response = broadcast_helper(&mut tx_builder, wallet).await;
+
+        while tx_retry {
+            tx_retry = false;
+
+            // On error, we verify it matches the retry strategies. If it does, we update the retry counts and start over
+            for s in self.strategies.iter_mut() {
+                if strategy_condition_met(s, &tx_response) && can_retry(s) {
```

Comment:
If this condition is met for multiple strategies then the tx can be re-broadcasted multiple times for one failure, is that the desired behavior?

Reply:
Nope, this is not true. The `continue` keyword ensures that it's retried only once (L96)

Reply:
Ahh okey so if the first strategy matches it only does that one? Should be documented on the constructor because that means that ordering matters.

Reply:
No if the strategy matches the first one, it tries and correct that one and then retries. If it then matches the second one, it retries with the second strategy in mind and so forth. Anyway, Node error messages only come one per one